src/github.com/matrix-org/dendrite/mediaapi/writers/upload.go

```diff
@@ -0,0 +1,244 @@
+			"inline; filename*=utf-8''" + url.PathEscape(string(r.MediaMetadata.UploadName)),
+		)
+	}
+
+	logger.WithFields(log.Fields{
+		"Origin":              r.MediaMetadata.Origin,
+		"UploadName":          r.MediaMetadata.UploadName,
+		"Content-Length":      r.MediaMetadata.ContentLength,
+		"Content-Type":        r.MediaMetadata.ContentType,
+		"Content-Disposition": r.MediaMetadata.ContentDisposition,
+	}).Info("Uploading file")
+
+	writer, file, tmpDir, errorResponse := createTempFileWriter(cfg.BasePath, logger)
+	if errorResponse != nil {
+		return *errorResponse
+	}
+	defer file.Close()
+
+	// The limited reader restricts how many bytes are read from the body to the specified maximum bytes
+	// Note: the golang HTTP server closes the request body
+	limitedBody := io.LimitReader(req.Body, int64(cfg.MaxFileSizeBytes))
+	hasher := sha256.New()
+	reader := io.TeeReader(limitedBody, hasher)
+
+	bytesWritten, err := io.Copy(writer, reader)
+	if err != nil {
+		logger.Warnf("Failed to copy %q\n", err)
+		removeDir(tmpDir, logger)
+		return util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.Unknown(fmt.Sprintf("Failed to upload")),
+		}
+	}
+
+	writer.Flush()
+
+	if bytesWritten != int64(r.MediaMetadata.ContentLength) {
+		logger.Warnf("Bytes uploaded (%v) != claimed Content-Length (%v)", bytesWritten, r.MediaMetadata.ContentLength)
+	}
+
+	hash := hasher.Sum(nil)
+	r.MediaMetadata.MediaID = types.MediaID(base64.URLEncoding.EncodeToString(hash[:]))
+
+	logger.WithFields(log.Fields{
+		"MediaID":             r.MediaMetadata.MediaID,
+		"Origin":              r.MediaMetadata.Origin,
+		"UploadName":          r.MediaMetadata.UploadName,
+		"Content-Length":      r.MediaMetadata.ContentLength,
+		"Content-Type":        r.MediaMetadata.ContentType,
+		"Content-Disposition": r.MediaMetadata.ContentDisposition,
+	}).Info("File uploaded")
+
+	// check if we already have a record of the media in our database and if so, we can remove the temporary directory
+	mediaMetadata, err := db.GetMediaMetadata(r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+	if err == nil {
+		r.MediaMetadata = mediaMetadata
+		removeDir(tmpDir, logger)
+		return util.JSONResponse{
+			Code: 200,
+			JSON: uploadResponse{
+				ContentURI: fmt.Sprintf("mxc://%s/%s", cfg.ServerName, r.MediaMetadata.MediaID),
+			},
+		}
+	} else if err != sql.ErrNoRows {
+		logger.Warnf("Failed to query database for %v: %q", r.MediaMetadata.MediaID, err)
+	}
+
+	// TODO: generate thumbnails
+
+	finalPath := getPathFromMediaMetadata(r.MediaMetadata, cfg.BasePath)
+
+	err = moveFile(
+		types.Path(path.Join(string(tmpDir), "content")),
+		types.Path(finalPath),
+	)
+	if err != nil {
+		logger.Warnf("Failed to move file to final destination: %q\n", err)
+		removeDir(tmpDir, logger)
+		return util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.Unknown(fmt.Sprintf("Failed to upload")),
+		}
+	}
+
+	err = db.StoreMediaMetadata(r.MediaMetadata)
```

Comment:
Am I missing some subtlety here? Why do we move to the destination and then insert the media metadata? It's annoying like this because if things fail we need to `rm` from different places.

I'm guessing it's possible that if we inserted metadata first it could race with a download request which would then not find the file on-disk, but we don't seem to note this anywhere.